Add CycloneDx output option

agschrei opened the pull request:
<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

Implements  #1888 

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

agschrei committed:
add scaffolding for cyclonedx output plugin #1888

introduce new --cyclonedx{-json} output options and register associated plugins in setup.cfg

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
extract CycloneDX components from codebase packages #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
move external reference mapping into separate function #1888

suppress keys with None values from output

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
refine mapping of declared_license fields to CycloneDx licenses #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
filter duplicate license entries from output #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
add custom JSONEncoder #1888

added custom JSONEncoder to rename fields in compliance with CycloneDx spec

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
merge components avoiding duplicates #1888

avoid duplicate component entries by merging two components with the same bom-ref

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
add indentation to make json output pretty-print #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
check in expected result for cyclonedx json output #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
add XML serialization #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
extract optional scope from dependency tree #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
docs: add cli-reference stub about new output options #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
test: verify removal of None or empty collections #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
handle scan results with empty packages #1888

do not serialize packages without name or version

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
CHANGELOG: add info about CycloneDx output plugin #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
test: update test_scan_cli_help test resource with new output options #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei commented:
Looking at the jobs that are currently failing it seems that the failures are related to the introduction of the `output_format_version` key in the JSON output which is independent of the changes this PR introduced.  
See for example [this job](https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=4623&view=logs&j=9a634e43-2dfe-534b-56d2-0b7ffad252eb&t=28d573b5-74f3-578a-960b-e035b88f9a42&l=4408)  

I'm not sure how to proceed here.

@pombredanne could you kindly provide some guidance?

JonoYang commented:
@agschrei Thank you for contributing a CycloneDX output plugin! We will review the plugin in depth, but it looks good so far. We are looking into the test error you mentioned, as it is affecting other branches as well.

pombredanne commented:
@agschrei Thank youn ++ for this... let me review.

pombredanne commented on tests/formattedcode/data/cyclonedx/expected.xml:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8"?><bom xmlns="http://cyclonedx.org/schema/bom/1.3" version="1" serialNumber="urn:uuid:fcd9dd28-d6ee-47b6-a368-b0df8a152cdd"><metadata><timestamp>2021-09-06T20:52:47Z</timestamp><tools><tool><vendor>nexB Inc.</vendor><name>scancode-toolkit</name><version>21.8.4</version></tool></tools></metadata><components><component type="library" bom-ref="pkg:npm/npm@2.13.5"> ...
```
Would it be possible to get a pretty-printed XML output by default?

agschrei replied:
I noticed that `lxml` was already a dependency in the requirements.txt anyway, so I refactored to use that instead. Has a couple of advantages over `xml`, the biggest one being a pretty-print option.

pombredanne commented on tests/formattedcode/data/cyclonedx/expected.json:
```diff
@@ -0,0 +1,122 @@
+      "purl": "pkg:npm/abbrev@1.0.7",
+      "licenses": [
+        {
+          "license": {
```
It looks like you are taking the declared license for licenses. This would be the license that's asserted in a package manifest, not detected and not normalized. You should instead use the license_expression. And likely convert this to an SPDX expression?

agschrei replied:
@pombredanne I already mentioned this a bit further down, but my reasoning was that the license expression inferred by scancode can sometimes be needlessly complex, see this example from the tests:
```json
 "purl": "pkg:npm/has@1.0.3",
      "licenses": [
        {
          "expression": "mit AND (mit AND (mit AND unknown))"
        },
        {
          "license": {
            "id": "MIT"
          }
        }
      ],
```

Granted this can be normalized to `mit AND unknown` but at that point it might still be helpful to have the declared license alongside it.

If license_expression and declared_license are one and the same, we only keep one of the entries, so this approach should never lead to duplicate information.

pombredanne replied:
> my reasoning was that the license expression inferred by scancode can sometimes be needlessly complex, see this example from the tests:

There is no such thing as a "needlessly complex expression". There are two cases:
1. an expression is correct and this should be the default case
2. an expression is not correct and this a bug

In the case of the `has` npm package this is a bug and we should treat this correctly 
Can you enter a bug?

https://github.com/tarruda/has/blob/8146666148f1812f90e737aeefcded17a8cf81f4/package.json#L24 

> If license_expression and declared_license are one and the same, we only keep one of the entries, so this approach should never lead to duplicate information.

As explained above the `declared_license` is not something that is a reliable data point as this has not gone through detection and normalization using the license detection engine, it should never be use as the license in a CycloneDX document.

pombredanne commented on tests/formattedcode/data/cyclonedx/expected.json:
```diff
@@ -0,0 +1,122 @@
+          "type": "bom"
+        }
+      ],
+      "type": "library",
```
Would not these field come further up, e.g. before the references?

agschrei replied:
I've moved these up so they no longer get lost behind the ext references which can get quite large

pombredanne commented on docs/source/rst_snippets/output_format_options.rst:
```diff
@@ -31,4 +31,8 @@ All Scan Output Options
 --html-app FILE         Write scan output as a mini HTML
                         application to FILE.
 
+--cyclonedx FILE        Write scan output as a CycloneDx 1.3 BOM in XML format to FILE
```
I wonder if it would make more sense to have `--cyclonedx` produce JSON and `--cyclonedx-xml` produce XML, e.g. promoting using JSON rather than XML?

agschrei replied:
I've switched these two flags like you suggested

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+from typing import FrozenSet, List, Tuple
+
+
+def _get_set_of_known_licenses_and_spdx_license_ids() -> Tuple[List, FrozenSet[str]]:
```
I am not against using type annotations per se, but for now we are not using them anywhere in ScanCode and until we are doing type checking in the CI and at runtime, what is the value of these beside documentation?

agschrei replied:
Depending on the IDE that you are developing in these can significantly improve autocompletion suggestions, so I have made a habit of using them.  
I've also noticed that the codebase doesn't make use of them, but `typing` was already part of the requirements.txt so I figured you are not directly opposed to adopting them.

pombredanne replied:
fair enough.  :+1:

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    """
+    from licensedcode.models import get_all_spdx_keys, load_licenses
+    licenses = load_licenses(with_deprecated=True)
+    spdx_keys = filter(lambda x: "LicenseRef" not in x,
```
Why do you filter LicenseRef here? that's pretty essential to license detection accuracy as you are putting aside ~1290 licenses out of 1879, e.g. you are discarding roughly over 65% of the licenses. Also you likely do not want to keep deprecated licenses `with_deprecated=True` above anyway.

agschrei replied:
this is now obsolete as I am getting the licenses from the cache (no longer filtering on LicenseRef prefix) as you suggested bellow.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    """load all licenses and all SPDX Ids known to ScanCode
+       this will also load scancode licenserefs, so we filter those
+    """
+    from licensedcode.models import get_all_spdx_keys, load_licenses
```
nitpicking: We prefer to have one import per line.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+def _get_set_of_known_licenses_and_spdx_license_ids() -> Tuple[List, FrozenSet[str]]:
+    """load all licenses and all SPDX Ids known to ScanCode
+       this will also load scancode licenserefs, so we filter those
+    """
```
nitpicking: We use rather this style for docstrings:
```
    """
    Return a tuple of (list of License, set of SPDX license ids).
    The SPDX license ids set does not include LicenseRef from ScanCode or other origins 
    """ 
```
key points:
- Start on the next line after """, Capitalize the first word.
- Use imperative style, and explain first what is returned

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+known_licenses, spdx_ids = _get_set_of_known_licenses_and_spdx_license_ids()
+
+hash_type_mapping = {
```
nitpicking: What about this instead:
```
cdx_hash_types_by_scancode_hash

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+@attr.s
+class CycloneDxExternalRef:
+    url: str = attr.ib()
+    type: str = attr.ib(validator=attr.validators.in_(["vcs", "issue-tracker", "website",
```
Nitpicking: for such list we have adopted the black style (but have not yet switched to it... and we still use single quotes)... so that would likely come out this way:
```Python
    type: str = attr.ib(validator=attr.validators.in_([
        'vcs',
        'issue-tracker',
        'website',
        'advisories',
        'bom',
        'mailing-list',
        'social',
        'chat',
        'documentation',
        'support',
        'distribution',
        'license',
        'build-meta',
        'build-system',
        'other'
    ]))
```

but then, it could make sense to declare this as a constant instead, may be like this:

```Python
    reference_types = frozenset([
        'vcs',
        'issue-tracker',
        'website',
        'advisories',
        'bom',
        'mailing-list',
        'social',
        'chat',
        'documentation',
        'support',
        'distribution',
        'license',
        'build-meta',
        'build-system',
        'other'
    ])

    type: str = attr.ib(validator=attr.validators.in_(reference_types))

```

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+}
+
+
+def get_external_refs(package: dict) -> List[CycloneDxExternalRef]:
```
Could this be a class method of CycloneDxExternalRef, since this is essentially a `from_package()` factory?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    return ext_refs
+
+
+def get_hashes_list(package: dict) -> List[CycloneDxHashObject]:
```
Same as above, could this be a class method of CycloneDxHashObject, since this is essentially a from_package() factory?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+def get_author_from_parties(parties: List[dict]) -> str:
+    if parties is not None:
+        try:
+            # assume there can only ever be one author and get that with next()
```
This assumption does not hold. There is no bound on the number of authors in a parties list. I wish it were, but packages models are all over the map data-structure wise?

agschrei replied:
I assumed so. For npm, as an example, many projects list multiple authors.  
The tricky bit here is that cdx expects a string as the value for the 'author' field.  
My guess is that cdx expects there to only ever be one author, hence referring to the singular.  

If we want to map multiple authors to this field, how would you like to proceed?

My suggestion is mapping authors in a CSV-like fashion:
`"author": " 'author a', 'author b' "`
Do you see a more elegant soluton here? I'm not happy with this, but it beats leaving out authors in the mapping.

pombredanne replied:
I would guess that a better approach would be to list all authors indeed, and just just their plain names, one per line

agschrei replied:
that's how we are doing it now to ensure all authors are considered

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+            return None
+
+
+"""
```
May be this should be the docstring of the CycloneDxOutput class instead?

agschrei replied:
true, accidentally separated this from the class

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+        try:
+            # assume there can only ever be one author and get that with next()
+            author = next(filter(lambda p: p.get("role") == "author", parties))
+            return ' '.join(filter(None, (author.get("name"),
```
nit: let's use black wrapping style

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+@output_impl
+class CycloneDxOutput(OutputPlugin):
+    options = [
+        PluggableCommandLineOption(('--cyclonedx', 'output_cyclonedx',),
```
nit: let's use black wrapping style

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+@output_impl
+class CycloneDxJsonOutput(OutputPlugin):
+    options = [
+        PluggableCommandLineOption(('--cyclonedx-json', 'output_cyclonedx_json',),
```
nit: let's use black wrapping style

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+
+def merge_components(existing: CycloneDxComponent, new: CycloneDxComponent):
+    """merges two components and returns a new component"""
```
nit: use our docstring style

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    )
+
+
+def merge_components(existing: CycloneDxComponent, new: CycloneDxComponent):
```
What is the purpose of merging? can this happen between two packages that have different purl? should this be checked?

agschrei replied:
Since we use the `purl` as a `bom-ref` it has to be unique, merging boms that have the same purl ensures this.  
This method now checks if the `purl` values are indeed identical before merging and raises a `ValueError` if they are not.

Please note that the only location where we call this helper doesn't catch that because the context in which it is invoked already ensures that both components have the same purl

pombredanne replied:
This may be a problem. ScanCode reports all instances of each package so there can be duplicated purls (and this is evolving further with https://github.com/nexB/scancode-toolkit/pull/2748 )
The thing is that each package instance may have the same purl but can different attributes (for instanced when derived from a certain manifest vs. another format of style such as a Gemfile, gemspec and installed metadata.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+
+def build_bom(codebase) -> CycloneDxBom:
+    # TODO: find out if we can always expect that header to be present
```
It is guaranteed to be always there, but there could be multiple ones( it is a list), possibly each list item from a different tool.

agschrei replied:
the approach that I am taking here:
I'm filtering the headers for the `scancode-toolkit` header and only forwarding that to the CycloneDx output.  Is this an acceptable solution for you?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    files = OutputPlugin.get_files(codebase)
+
+    # get all packages that are not None
+    packages = [package for file in files for package in file.get("packages")]
```
nit: for consistency throughout, use single quotes for now everywhere except for docstrings

agschrei replied:
this is now enforced across all strings

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+    components = generate_component_list(packages)
+    # associate components by purl
+    comp_map = dict(map(lambda c: (c.purl, c), components))
```
what about naming this `components_by_purl`, this means no comment is needed.
Also what if there are several times the same purl for different packages instances?

agschrei replied:
I've renamed this to be more expressive of its intent.  
At this point thanks to merging of components with the same purl we are already guaranteed to only have one component per purl.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+        merge_lists(existing.properties, new.properties)
+
+
+def generate_component_list(packages) -> List[CycloneDxComponent]:
```
could this be a class factory method of CycloneDxComponent instead? may be named `from_packages()`?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+
+def generate_component_list(packages) -> List[CycloneDxComponent]:
+    ref_component_map = {}
```
nit: rather than suffix a name with its type (e.g. `_map`) we prefer to use a more descriptive name such as `components_by_purl`

agschrei replied:
taken into consideration not just here but for all members

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    json.dump(bom, output_file, indent=2, sort_keys=False, cls=CycloneDxEncoder)
+
+
+class XmlSerializer():
```
why the trailing `()`?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+        return bom_element
+
+    def _get_tool_element(self) -> ET.Element:
+        tool = ET.Element("tool")
```
nit: let's use single quotes consistently

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    return None
+
+
+def get_license_entry_from_license_expression(license_expression: str) \
```
nit: please avoid using backslash for line continuations. Wrapping black-style is best. 90'sih line length is OK too.
Also, could this be instead a class  method on the CycloneDxLicenseEntry class ?

agschrei replied:
fixed this

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    return (licenses, frozenset(spdx_keys))
+
+
+known_licenses, spdx_ids = _get_set_of_known_licenses_and_spdx_license_ids()
```
IMHO, avoid to make such long running call at an import time. 

Use the ` licensedcode.cache` module instead to get what you need with `get_licenses_db()` to get a mapping of `{license key: License object}`

agschrei replied:
not sure if I understood you correctly, but I am now getting the licenses from the cache and extracting the spdx IDs from there

pombredanne replied:
My point was also to move this line into a function and not call the function at import  time

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+            lic.url = declared_license
+        else:
+            lic.name = declared_license
+    # some declared_license entries are also expressed as dicts
```
.... and some as list or anything. By construction the data structure is whatever existed upstream. I think you should not use the declared license in the output but instead only the `license_expression` which is the detection that took place

agschrei replied:
I was going for the low-hanging fruit here as we probably can't cover everything.  
But as raised [here](https://github.com/nexB/scancode-toolkit/pull/2698#discussion_r737205459) I'm currently keeping declared licenses around in cases where the detection differs from the declared license

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    else:
+        licenses = []
+
+    declared_license = package["declared_license"]
```
Do not use both declared_license and license expression, as this is redundant.

agschrei replied:
I have a question about this:

In one of the larger tests I ran the `license_expression` produced by scancode and the declared package license looked like this:

```json
 "purl": "pkg:npm/has@1.0.3",
      "licenses": [
        {
          "expression": "mit AND (mit AND (mit AND unknown))"
        },
        {
          "license": {
            "id": "MIT"
          }
        }
      ],
```

I know too little about your license detection engine to explain where this detection comes from, but in that case having the declared license alongside the expression can be helpful.  

The approach I am currently taking here is:
- add the license expression as a license entry
- only add declared license(s) if they differ from the license expression

Given the example above, is this a viable option?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    return CycloneDxLicenseEntry(license=lic, expression=None)
+
+
+def get_licenses(package: dict) -> List[CycloneDxLicenseEntry]:
```
Could this be a factory class method on CycloneDxLicenseEntry instead?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+    license = known_licenses.get(license_expression)
+    if license is not None:
+        # attempt to set either official OSI URL or any license text URL
```
IMHO a better URL is https://github.com/nexB/scancode-toolkit/blob/develop/src/scancode/api.py#L141

agschrei replied:
I didn't know about this one, but we are now using it instead of the OSI URL

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+    return obj_dict
+
+
+class CycloneDxEncoder(json.JSONEncoder):
```
Could we instead have an `to_dict()` methods on objects to serialize and avoid having JSON-specific serialization code?

agschrei replied:
@pombredanne I am open to this but would like to understand your reasoning about it first. Is there some downside to the serialization code that I am not seeing?  

I solved it this way because this encoder throws away any fields with empty or None values and renames the `bom_ref` field to `bom-ref` for output no matter what we throw in, so it works generically for all objects we work with here

pombredanne replied:
> I am open to this but would like to understand your reasoning about it first. Is there some downside to the serialization code that I am not seeing?

We also use a `to_dict()` function throughout the codebase as a convention. 
A JSONEncoder makes the serialization JSON-specific and impossible to use with other presentations such as YAML and I would rather have an explicit field-by-field mapping rather than something implicit. 
Also why would you rename fields?  Does the format demands to have "bom-ref" vs. "bom_ref" ?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+class CycloneDxEncoder(json.JSONEncoder):
+    """Custom encoder that removes fields which are either empty or None.
+       Additionally renames the `bom_ref` field to `bom-ref` for output"""
+
```
Is this the standard for CycloneDX to skip empty fields?

agschrei replied:
I'm not sure if CycloneDx defines a preference on that, but intuitively it made sense for me to prune anything that is either empty or None as such entries add no value to the BOM

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+
+@attr.s
+class CycloneDxHashObject():
+    alg: str = attr.ib()
```
may be using a full name doe not hurt, like algorithm? unless this is the cdx name of course.

agschrei replied:
this is the cdx name, so I'd petition to keep it like this so that the default __dict__ repr returns what we want the JSON key to be without modification

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+        url = license.osi_url
+        if license.text_urls:
+            url = license.text_urls[0]
+        lic = CycloneDxLicense(id=license.spdx_license_key,
```
Do you need an SPDX license expression? if so we can get this directly instead.

What is CycloneDxLicenseEntry? a single license or an expression?

agschrei replied:
I am currently using the scancode license expression here as is, because the BOM format doesn't specify anything further about the expression field. But if we can get an SPDX license expression from that directly I would be in favor of that. Where do I need to look? Does the codebase offer some kind of utility to convert to that?

A `CycloneDxLicenseEntry` is a union type that can either hold a single expression or a  `CycloneDxLicense` that has an (spdx) id, name and optional URL. CycloneDx also allows to include the license text here, but we currently don't map this as I believe including the license text for every finding would only needlessly bloat the report.  
My argument here is that if a user / downstream tool wants the license text they can just grab it from the provided URL

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,659 @@
+            digest = package[alg]
+            if alg is not None and digest is not None:
+                hashes.append(CycloneDxHashObject(
+                    hash_type_mapping[alg], digest))
```
What about this instead? Also note how I use named keyword arguments which helps if we ever go through some name refactoring and avoid the order dependency implied by positional arguments.

```
    for scancode_algo, cyclonedx_algo in hash_type_mapping.items():
        digest = package.get(scancode_algo)
        if digest:
            hashes.append(CycloneDxHashObject(algo=cyclonedx_algo, content=digest))
```

agschrei replied:
that does look a lot cleaner, incorporated this into the changes

pombredanne reviewed (changes requested):
This is a beautiful PR!
I came with a few nitpickings for your consideration. One thing I am not sure I get is how licenses or license expressions are stored in CycloneDX.
We can have also some chat at your convenience; we meet at https://gitter.im/aboutcode-org/discuss

agschrei commented:
> This is a beautiful PR!  

Thanks, I have to say that your contribution guide is really helpful for first-time contributors.  

> I came with a few nitpickings for your consideration.

I'll have to set some time aside to implement these changes.

Looking at your comments most of the open points should be quick fixes,  
but there's some other stuff like the declared licenses topic that might require additional back and forth.

So there's a good chance I'll try to find you on gitter like you suggested to clarify on that.

> One thing I am not sure I get is how licenses or license expressions are stored in CycloneDX.

The way I understood the CycloneDx spec, a component may have a list of `License` entries where each entry may either have    
- a license expression as a string
- a `License Object` as the spec calls it   

but obviously not both.  
So I tried to implement that as an explicit union type with `CycloneDxLicenseEntry`.


I used Atlassian's great schema viewer to explore the CycloneDx 1.3 schema [here](https://json-schema.app/view/%23/%23%2Fdefinitions%2Fcomponent/%23%2Fdefinitions%2FlicenseChoice?url=https%3A%2F%2Fraw.githubusercontent.com%2FCycloneDX%2Fspecification%2Fmaster%2Fschema%2Fbom-1.3.schema.json) if you're interested.

agschrei committed:
replace xml with lxml to allow easy pretty-printing #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
rename cyclonedx cli options to favor json over xml #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
reorder JSON keys and align with project's coding style #1888

put bom-ref, type and scope keys higher up in component entries,
consistently use single quotes

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

pombredanne commented:
> The way I understood the CycloneDx spec, a component may have a list of `License` entries where each entry may either have
> 
>     * a license expression as a string
> 
>     * a `License Object` as the spec calls it
> 
> 
> but obviously not both.
> So I tried to implement that as an explicit union type with `CycloneDxLicenseEntry`.
> 
> I used Atlassian's great schema viewer to explore the CycloneDx 1.3 schema [here](https://json-schema.app/view/%23/%23%2Fdefinitions%2Fcomponent/%23%2Fdefinitions%2FlicenseChoice?url=https%3A%2F%2Fraw.githubusercontent.com%2FCycloneDX%2Fspecification%2Fmaster%2Fschema%2Fbom-1.3.schema.json) if you're interested.

It makes sense. So I think that we should instead likely:
1. use only the license_expression field, as a bona fide SPDX expression
2. in order to also provide information about all the LicenseRef-scancode which are numerous, we could provide some 'license' externalReference for each non-SPDX license (or each license) as explained in https://json-schema.app/view/%23/%23%2Fdefinitions%2Fcomponent/%23%2Fdefinitions%2FexternalReference?url=https%3A%2F%2Fraw.githubusercontent.com%2FCycloneDX%2Fspecification%2Fmaster%2Fschema%2Fbom-1.3.schema.json
3. we could consider also keep the matched license notice text somewhere in the reference comment, but I do not think this can work as each license match may have a different notice even when matched to the same expression

agschrei committed:
apply black formatting without string normalization #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
fix failing test by updating expected result #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

pombredanne commented:
@agschrei thanks for the updates. Let me review.

agschrei committed:
rename variables and reliably extract scancode-toolkit header #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

pombredanne commented:
@agschrei gentle ping. I would like to merge this soon enough.

agschrei committed:
refactor: turn standalone functions into class methods #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei committed:
map multiple authors and use license db url as fallback #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei commented:
Most of the requested changes are now implemented in the way I understood them.  
Please let me know if something still irks you about the modifications.

The only major thing that is still missing is the correct handling of `LicenseRef-*` SPDX IDs as suggested above, I will hopefully finish that tomorrow.  

I'll reach out to you on the gitter channel to double-check my solution then.

agschrei committed:
add external refs for scancode spdx ids #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

pombredanne commented:
I triggered a re-run of the test failures which are due to Azure flakiness, not a real error.

pombredanne commented:
and there is a new lint/check that was merged recently for the documentation and you are failing it... that's an easy fix though:

> Run ./scripts/doc8_style_check.sh
source/rst_snippets/output_format_options.rst:34: D001 Line too long
source/cli-reference/output-format.rst:585: D001 Line too long
Error: Process completed with exit code 1.

agschrei committed:
docs: break lines at 100 chars length to pass lint check #1888

Signed-off-by: Alexander Gschrei <alexander.gschrei@fau.de>

agschrei commented:
I just fixed the docs sections that were failing the lint check.  

Hopefully everything passes now.

pombredanne commented:
@agschrei gentle ping... there are few nitpickings left for your review

agschrei commented:
@pombredanne I was still waiting on feedback for some of the points you raised. I have now resolved everything that is fixed/considered, but for the points that still remain open I either need additional input or just a nod that this is how you want to proceed. Could you please have a look at the comments there?

Also, should I rebase this onto develop in the meantime to see if it plays nice with the latest changes there?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+
+
+known_licenses, spdx_ids = _get_set_of_known_licenses_and_spdx_license_ids()
+spdx_licenseref_prefix = "LicenseRef-scancode-"
```
You should get the SPDX license key instead from the license records. This "LicenseRef-scancode-" cannot be used safely as the set of licenses evolves continuously

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+@attr.s
+class CycloneDxLicenseEntry:
+    @classmethod
+    def from_declared_license(cls, declared_license) -> 'CycloneDxLicenseEntry':
```
You cannot use the "declared license" safely: this is a data structure which can be anything and is whatever was upstream in the package manifest. It could be a list, string, mapping... anything declared upstream.
You want instead to use the "license_expression" (and likely the SPDX license expression).

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+
+        from scancode.api import SCANCODE_LICENSEDB_URL
+
+        license = known_licenses.get(license_expression)
```
This will rarely work, as a license expression is not a license, but eventually multiple licenses. The license_expression would need to be broken down in its keys (using the `license_expression` library`) and  each key looked up in the license db.
Is there a way to store an SPDX expression directly in the CycloneDX doc instead

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+    def from_package(cls, package: dict) -> List['CycloneDxLicenseEntry']:
+        """
+        Return a list of CycloneDxLicenseEntry instances built from package data
+        Attempt to map `license_expression` and `declared_license` data to CycloneDx format
```
As mentioned elsewhere, do not use `declared_license`, use only `license_expression`.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+        Attempt to map `license_expression` and `declared_license` data to CycloneDx format
+        """
+        # store previously encountered IDs to identify duplicates
+        seen_ids = set()
```
You cannot deduplicate based on license ids. Licenses are in an expression that can be logically simplified, and logically deduplicated, but doing it on the keys of an expression could lead to this issues: say the expression is: `(MIT OR GPL-2.0-only) AND GPL-2.0-only` ... your approach would likely damage the meaning of the original expression. You cannot process license ids when expressions are used throughout here or you risk returning incorrect or misleading results.

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+
+        return licenses
+
+    license: CycloneDxLicense = attr.ib(default=None)
```
Please move attributes declaration at the top of the class

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+
+@attr.s
+class CycloneDxMetadata:
+    timestamp = attr.ib(default=datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
```
This is executed at import time and is a global: all the object will shared the same value. This is not what you want ... you need a factory https://www.attrs.org/en/stable/api.html?highlight=factory#attr.Factory

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+@attr.s
+class CycloneDxMetadata:
+    timestamp = attr.ib(default=datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
+    tools: List[dict] = attr.ib(factory=list)
```
list is likely not what you want: See https://www.attrs.org/en/stable/api.html?highlight=factory#attr.Factory

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+
+@attr.s
+class CycloneDxHashObject:
+    __cdx_hash_types_by_scancode_hash = {
```
Why the __ prefix?

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+            if digest:
+                hashes.append(CycloneDxHashObject(alg=cyclonedx_alg, content=digest))
+        return hashes
+    alg: str = attr.ib()
```
Please put attributes at the top

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+        licenses = filter(None, map(lambda lic: lic.license, self.licenses))
+        license_ids = map(lambda lic: lic.id, licenses)
+        licenseref_ids = filter(
+            lambda lic_id: lic_id is not None and lic_id.startswith("LicenseRef-scancode-"),
```
As example elsewhere, you cannot craft SPDX ids this way. You need to grab the stored spdx_license_key of a license

pombredanne commented on src/formattedcode/output_cyclonedx.py:
```diff
@@ -0,0 +1,774 @@
+        merge_lists(existing.properties, new.properties)
+
+
+def truncate_none_or_empty_values(obj) -> dict:
```
Why do you need this? Could you also have some tests for this?

pombredanne reviewed (changes requested):
Sorry for the late review!
See my comments inline for your consideration.

pombredanne commented:
> my reasoning was that the license expression inferred by scancode can sometimes be needlessly complex, see this example from the tests:

Note that we are working on introducing a "primary license" for packages... which would address your concern. In the meantime, I will merge your branch patched to use the `license_expression` and not the `declared_license`.

pombredanne committed:
Merge latest develop branch

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Sort and streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new build_spdx_license_expression() function

This new function returns an SPDX license expression string from a
ScanCode license_expression string.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine CycloneDX support #1888

Only deal with an SPDX license_expression and not plain licenses or
declared licenses, because this is the only correct license information.

Removing most type hints since we are not yet checking these.

Remove dependency lookup which could be misleading. Instead add warning
if dependencies are skipped.

Add docstring to most methods and functions.

Streamline nested list comprehensions and filter/map constructions and
replace these with simpler constructs.

Replace JSON encoder by a plain to_dict() serialization throughout.

Move attrs attributes declarations towards the top of classes
declarations rather than at the bottom.

Move move method-like functions to their respective objects, including
moving XML serialization functions to to_xml_element() methods for each
model object.

Use factories for fields defaults such as URN and timestamp.

Split method that work on ScanCode packages lists in two: one operating
on a single package mapping and one working on the packages list.

Update tests to work on larger set of expected data.

Always return all attributes, including empty values for now.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
@agschrei I merged the latest develop and updated your code according to some of the comments above. Please review before I merge!
Thank you for your patience.

The key changes are outlined in the commit message of https://github.com/nexB/scancode-toolkit/pull/2698/commits/953a40edd60551477986150e71083e09fce407e0

## Refine CycloneDX support #1888

- Only deal with an SPDX license_expression and not plain licenses or
declared licenses, because this is the only correct license information.

- Removing most type hints since we are not yet checking these.

- Remove dependency lookup which could be misleading. Instead add warning
if dependencies are skipped.

- Add docstring to most methods and functions.

- Streamline nested list comprehensions and filter/map constructions and
replace these with simpler constructs.

- Replace JSON encoder by a plain to_dict() serialization throughout.

- Move attrs attributes declarations towards the top of classes
declarations rather than at the bottom.

- Move move method-like functions to their respective objects, including
moving XML serialization functions to to_xml_element() methods for each
model object.

- Use factories for fields defaults such as URN and timestamp.

- Split method that work on ScanCode packages lists in two: one operating
on a single package mapping and one working on the packages list.

- Update tests to work on larger set of expected data.

- Always return all attributes, including empty values for now.

pombredanne committed:
Revert change in SDPX tests XML loading

The ways of XML and encoding are sometimes complex.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
All green now. Merging!
